source/unix/openssl_rsa.c

```diff
@@ -0,0 +1,375 @@
+
+        if (s_reinterpret_evp_error_as_crt(
+            EVP_PKEY_CTX_set_signature_md(ctx, EVP_sha256()), "EVP_PKEY_CTX_set_signature_md")) {
+            return AWS_OP_ERR;
+        }
+    } else {
+        return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_rsa_sign(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signature_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_buf *out) {
+    struct lc_rsa_key_pair *key_pair_impl = key_pair->impl;
+
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(key_pair_impl->key, NULL);
+    if (ctx == NULL) {
+        return aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+    }
+
+    if (s_reinterpret_evp_error_as_crt(
+        EVP_PKEY_sign_init(ctx), "EVP_PKEY_sign_init")) {
+        goto on_error;
+    }
+
+    if (s_set_signature_ctx_from_algo(ctx, algorithm)) {
+        goto on_error;
+    }
+
+    size_t ct_len = out->capacity - out->len;
+    if (s_reinterpret_evp_error_as_crt(
+        EVP_PKEY_sign(ctx, out->buffer, &ct_len, digest.ptr, digest.len), "EVP_PKEY_sign")) {
+        goto on_error;
+    }
+    out->len += ct_len;
+
+    EVP_PKEY_CTX_free(ctx);
+    return AWS_OP_SUCCESS;
+
+on_error:
+    EVP_PKEY_CTX_free(ctx);
+    return AWS_OP_ERR;
+}
+
+static int s_rsa_verify(
+    const struct aws_rsa_key_pair *key_pair,
+    enum aws_rsa_signature_algorithm algorithm,
+    struct aws_byte_cursor digest,
+    struct aws_byte_cursor signature) {
+    struct lc_rsa_key_pair *key_pair_impl = key_pair->impl;
+
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(key_pair_impl->key, NULL);
+    if (ctx == NULL) {
+        return aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+    }
+
+    if (s_reinterpret_evp_error_as_crt(
+        EVP_PKEY_verify_init(ctx), "EVP_PKEY_verify_init")) {
+        goto on_error;
+    }
+
+    if (s_set_signature_ctx_from_algo(ctx, algorithm)) {
+        goto on_error;
+    }
+
+    int error_code = EVP_PKEY_verify(ctx, signature.ptr, signature.len, digest.ptr, digest.len);
+    EVP_PKEY_CTX_free(ctx);
+
+    /*
+    * Verify errors slightly differently from the rest of evp functions. 0
+    * indicates signature does not pass verification, and negative indicates
+    * that something went wrong. Manually handle error instead of using helper.
+    */
+    if (error_code > 0) {
+        return AWS_OP_SUCCESS;
+    } else { 
+
+        uint32_t error = ERR_peek_error();
+        AWS_LOGF_ERROR(AWS_LS_CAL_RSA, "Calling function %s failed with %d and extended error %lu",
+            "EVP_PKEY_verify", error_code, (unsigned long)error);
+
+        if (error == 0) {
+            return aws_raise_error(AWS_ERROR_CAL_SIGNATURE_VALIDATION_FAILED);
+        } else if (error == -2) {
+            return aws_raise_error(AWS_ERROR_CAL_UNSUPPORTED_ALGORITHM);
+        } else {
+            return aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        }
+    }
```

Comment:
trivial: we can still use the helper
```suggestion
    /* Verify errors slightly differently from the rest of evp functions.
     * 0 indicates signature does not pass verification, it's not necessarily an error. */
    if (error_code > 0) {
        return AWS_OP_SUCCESS;
    } else if (error_code == 0) {
       return aws_raise_error(AWS_ERROR_CAL_SIGNATURE_VALIDATION_FAILED);
    } else {
        return s_reinterpret_evp_error_as_crt(error_code, "EVP_PKEY_verify");
    }
```